### ISSUE TITLE
Shipping Mockito Javadocs

### ISSUE THREAD

**Issue by mstachniuk** [mockito/mockito#1651]
> Hey,
> 
> First thanks for reporting, in order to help us to classify issue can you make sure the following check boxes are checked ?
> 
> If this is about mockito usage, the better way is to reach out to
> 
>  - stackoverflow : http://stackoverflow.com/questions/tagged/mockito
>  - the mailing-list  : https://groups.google.com/forum/#!forum/mockito / mockito@googlegroups.com
>    (Note mailing-list is moderated to avoid spam)
>
> _This block can be removed_
> _Something wrong in the template fix it here `.github/ISSUE_TEMPLATE.md`

Currently, Mockito Javadocs are published by external javadoc.io service: https://static.javadoc.io/org.mockito/mockito-core/2.24.5/org/mockito/Mockito.html
When https://github.com/mockito/shipkit/issues/233 will be implemented then Mockito Javadocs can be automatically published to git repository during release build. 
The question is, where publish Mockito javadocs? There are a few options:

1. repo: mockito/mockito branch: gh-pages
Currently, it redirects to javadoc.io only

2. repo: mockito/mockito branch: master
To keep Javadocs together with the code

3. repo: mockito/mockito.github.io branch: master directory: javadoc
In this case, Javadocs will be available automatically on mockito.org

4. repo: mockito/mockito-javadoc branch: master
Separate repository for Javadocs only. The repo needs to be created first by Mockito owners. This will be default in Shipkit.

**Comment by TimvdLippe** [comment]
> 1. repo: mockito/mockito branch: gh-pages
>    Currently, it redirects to javadoc.io only

I would like to keep the docs separate, preferably in mockito.github.io
> 2. repo: mockito/mockito branch: master
>    To keep Javadocs together with the code

This option is a no for me. It will clutter the diffs and will generate merge conflicts.
> 3. repo: mockito/mockito.github.io branch: master directory: javadoc
>    In this case, Javadocs will be available automatically on mockito.org

This is fine by me, but is there any reason we need to migrate off javadoc.io? What is javadoc.io providing us that we potentially can't do ourselves? We discussed this back in #692. Are these arguments still true?
> 4. repo: mockito/mockito-javadoc branch: master
>    Separate repository for Javadocs only. The repo needs to be created first by Mockito owners. This will be default in Shipkit.

What is the benefit of this option over just using mockito.github.io?

TLDR: Options 1 and 2 are a no for me. Option 4 is reasonable, but I don't see any strict benefits over option 3. I would like to see us discuss option 3 versus hosting ourselves.

**Comment by mstachniuk** [comment]
From #692 the point:
> there is a delay in showing docs after release

is still valid (at least after read main page of javadoc.io).
I made a test on javadoc.io (by manipulate URL) on a fresh library and I got: "Javadoc is being downloaded. Normally this will be completed within 10 minutes. Please refresh page again later."

Regarding:
> no way to link to specific class of latest release

is still valid (I didn't find a way to do that).

Task implemented in Shipkit will copy Javadoc to `/1.2.3` and `/current` directories, so Javadocs will be available without delay and link to specific class of latest release should be possible.

**Comment by bric3** [comment]
>> there is a delay in showing docs after release
>
> is still valid (at least after read main page of javadoc.io).

I believe this delay is something we can live with. 

>> no way to link to specific class of latest release
>
> is still valid (I didn't find a way to do that).

There : 
* https://www.javadoc.io/doc/org.mockito/mockito-core/2.24.0
* https://static.javadoc.io/org.mockito/mockito-core/2.24.0/org/mockito/Mockito.html


I'm not keen to have a continuously growing code repository just for javadocs with our release model, (release as we merge).
However using the website could be an option, that's what the infamous netty project is doing https://github.com/netty/netty.github.com, and many others. But javadoc.io liberates us from the maintenance cost, none, which is a strong selling point !

**Comment by mockitoguy** [comment]
I agree that the delay or inconvenient links is something we could live with. However, there is a deal breaker with javadoc.io. We don't have javadoc for our "minor" releases, only for those that are in Maven Central.

The implementation will be in Shipkit so I don't see a direct maintenance cost on Mockito. If it stops working... well, there is still javadoc.io :) I don't see any downsides for Mockito provided that we use a separate repository for javadoc. Let's do it!

**Comment by mockitoguy** [comment]
Let's go for option 4 because this ensures that repos are neatly decoupled. Re-using the same repo for source code and for hosting is not a good idea (size, history).

**Comment by mockitoguy** [comment]
We already have Javadoc repo: https://github.com/mockito/javadoc/, @mstachniuk, can you use it for testing?

**Comment by TimvdLippe** [comment]
> Let's go for option 4 because this ensures that repos are neatly decoupled. Re-using the same repo for source code and for hosting is not a good idea (size, history).

We already have site.mockito.org which has the link to the Javadoc. Are you planning on having a separate domain for that? If not, I think we should put it on mockito.org/javadoc instead (e.g. sub-directory of github.com/mockito/mockito.github.io)

**Comment by TimvdLippe** [comment]
> I don't see any downsides for Mockito provided that we use a separate repository for javadoc.

It's is mostly server maintenance. We somehow need to serve the content. For GitHub pages, the *.github.io repository is the way to serve content from a GitHub repository. If you have a separate repository (e.g. mockito/javadoc), you will need to set it up yourself and configure/maintain these servers yourself. I don't think that is a viable option.

**Comment by mstachniuk** [comment]
Sure, I can use https://github.com/mockito/javadoc/ first for testing (after testing in Shipkit of course). 
Then you can decide to stay on that repo or move Javadoc publication to mockito.github.io

@mockitoguy please check Shipkit Slack ;-)

**Comment by mockitoguy** [comment]
> If you have a separate repository (e.g. mockito/javadoc), you will need to set it up yourself and configure/maintain these servers yourself.

@TimvdLippe, we already have mockito/javadoc repo that serves from GitHub at http://site.mockito.org/javadoc/current/

**Comment by TimvdLippe** [comment]
@mockitoguy I was not aware of us shipping our Javadoc on our own, especially as we link to the external documentation instead.

If that works without extra work, then sure. I don't have strong feelings about not shipping code to the repo either.
